File: python/64. minimum-path-sum.py
from typing import List


# lc sol start
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        n_rows, n_cols = len(grid), len(grid[0])

        for row in range(1, n_rows):
            grid[row][0] += grid[row - 1][0]

        for col in range(1, n_cols):
            grid[0][col] += grid[0][col - 1]

        for row in range(1, n_rows):
            for col in range(1, n_cols):
                grid[row][col] += min(grid[row - 1][col], grid[row][col - 1])

        return grid[-1][-1]

# lc sol end


def main() -> None:
    test_cases = {
        'inputs': [([[1, 3, 1], [1, 5, 1], [4, 2, 1]],), ([[1, 2, 3], [4, 5, 6]],)],
        'outputs': [7, 12],
    }

    solution = Solution()

    for idx, args in enumerate(test_cases['inputs']):
        op = solution.minPathSum(*args)
        assert op == test_cases['outputs'][idx], \
            f"expected output {test_cases['outputs'][idx]} for input {args}, got {op}"

    print('all sample test cases passed successfully!')


if __name__ == '__main__':
    main()

File: cpp/64. minimum-path-sum.cpp
//
// Created by Rutvik Patel on 1/31/24.
//

#include "include.h"

// solution
class Solution {
public:
    int minPathSum(std::vector<std::vector<int>>& grid) {
        int n_rows = static_cast<int>(grid.size()), n_cols = static_cast<int>(grid[0].size());
        std::vector<std::vector<int>> dp (n_rows, std::vector<int> (n_cols, std::numeric_limits<int>::max()));
        dp[0][0] = grid[0][0];

        for (int r_idx = 1; r_idx < n_rows; ++r_idx) dp[r_idx][0] = dp[r_idx - 1][0] + grid[r_idx][0];
        for (int c_idx = 1; c_idx < n_cols; ++c_idx) dp[0][c_idx] = dp[0][c_idx - 1] + grid[0][c_idx];

        for (int r_idx = 1; r_idx < n_rows; ++r_idx)
            for (int c_idx = 1; c_idx < n_cols; ++c_idx)
                dp[r_idx][c_idx] = grid[r_idx][c_idx] + std::min(dp[r_idx - 1][c_idx], dp[r_idx][c_idx - 1]);

        return dp.back().back();
    }
};

// main
int main(__attribute__((unused)) int argc, __attribute__((unused)) char *argv[]) {
    std::vector<std::pair<std::vector<std::vector<int>>, int>> test_cases = {
        { { { 1, 3, 1 }, { 1, 5, 1 }, { 4, 2, 1 } }, 7 },
        { { { 1, 2, 3 }, { 4, 5, 6 } }, 12 },
    };

    Solution solution;

    for (auto& [args, ans] : test_cases)
        assert_m(solution.minPathSum(args) == ans, "Failed!");

    log("All testcases passed!");
}